src/coprocessor/codec/mysql/duration.rs

```diff
@@ -13,215 +13,336 @@
+            let padding = round * 10;
+            nano = (nano / padding + 1) * padding;
+        }
+
+        second += nano / NANOS_PER_SEC;
+        minute += second / SECS_PER_MINUTE;
+        hour += minute / MINUTES_PER_HOUR;
+        check_hour(hour)?;
+
+        nano %= NANOS_PER_SEC;
+        second %= SECS_PER_MINUTE;
+        minute %= MINUTES_PER_HOUR;
 
-        let dur = StdDuration::new(
-            (nanos / NANOS_PER_SEC) as u64,
-            (nanos % NANOS_PER_SEC) as u32,
-        );
-        Duration::new(dur, neg, fsp)
+        let mut duration = Duration(0);
+        duration.set_neg(neg);
+        duration.set_hour(hour);
+        duration.set_minute(minute);
+        duration.set_second(second);
+        duration.set_nano(nano);
+        duration.set_fsp(fsp);
+        Ok(duration)
     }
 
-    pub fn new(dur: StdDuration, neg: bool, fsp: i8) -> Result<Duration> {
-        check_dur(&dur)?;
-        Ok(Duration {
-            dur,
-            neg,
-            fsp: check_fsp(fsp)?,
-        })
+    /// Try to build a Duration from `u64`
+    #[inline]
+    fn from_u64(inner: u64) -> Result<Duration> {
+        let duration = Duration(inner);
+        check_fsp(duration.fsp() as i8)?;
+        check_hour(duration.hour())?;
+        check_minute(duration.minute())?;
+        check_second(duration.second())?;
+        Ok(duration)
+    }
+
+    /// Try to build a `Duration` from `StdDuration` with `fsp` and `neg`
+    pub fn new(duration: StdDuration, neg: bool, fsp: i8) -> Result<Duration> {
+        let fsp = check_fsp(fsp)?;
+        let seconds = duration.as_secs();
+        let hour = check_hour(seconds / SECS_PER_HOUR)?;
+        let minute = seconds % SECS_PER_HOUR / SECS_PER_MINUTE;
+        let second = seconds % SECS_PER_MINUTE;
+        let nano = u64::from(duration.subsec_nanos());
+        Duration::with_detail(neg, hour, minute, second, nano, fsp)
     }
 
     /// Parses the time form a formatted string with a fractional seconds part,
     /// returns the duration type `Time` value.
     /// See: http://dev.mysql.com/doc/refman/5.7/en/fractional-seconds.html
-    pub fn parse(mut s: &[u8], fsp: i8) -> Result<Duration> {
+    pub fn parse(s: &[u8], fsp: i8) -> Result<Duration> {
         let fsp = check_fsp(fsp)?;
 
-        let (mut neg, mut day, mut frac) = (false, None, 0);
-
         if s.is_empty() {
             return Ok(Duration::zero());
-        } else if s[0] == b'-' {
-            s = &s[1..];
-            neg = true;
         }
 
-        let mut parts = s.splitn(2, |c| *c == b' ');
-        s = parts.next().unwrap();
-        if let Some(remain) = parts.next() {
-            let day_str = str::from_utf8(s)?;
-            day = Some(box_try!(u64::from_str_radix(day_str, 10)));
-            s = remain;
-        }
+        let mut s = str::from_utf8(s)?;
+        let origin = s;
+
+        // neg
+        let neg = if s.as_bytes()[0] == b'-' {
+            s = &s[1..];
```

Comment:
I guess that actually this operation could be slower when it is &str compared to &[u8] because &str additionally checks utf8 boundary: https://doc.rust-lang.org/std/primitive.str.html#method.get

Ditto for other string operations.

Reply:
Yes! Maybe we can just remove the boundary check because the following `parse::<u64>` will check the validity for us: 
```rust
let mut s = unsafe { str::from_utf8_unchecked(s) };
```

Reply:
@iosmanthus As it is `unchecked`, I think that line does not validate utf8. Maybe you can try use &[u8] first, to see how long it takes in parsing.

Reply:
If we do not transform the `s: &[u8]` into `s: &str`, we can not use method like `split/splitn`. Maybe we can rewrite the code as: 
```rust
let neg = if s[0] == b'-' {
    s = &s[1..];
    true
} else {
    false
};

let mut s = unsafe { str::from_utf8_unchecked(s) };
...
// somewhere we call parse::<u64> on s, if s contains invalid char(not a number)
// it will return an Err
```

Reply:
How about this one? https://doc.rust-lang.org/std/primitive.slice.html#method.splitn